src/api/firebase.js

```diff
@@ -92,16 +93,77 @@ export async function addItem(listId, { itemName, daysUntilNextPurchase }) {
 export async function updateItem(listId, itemId, checked) {
 	const itemRef = doc(db, listId, itemId);
 	const itemDoc = await getDoc(itemRef);
-	const { dateLastPurchased, totalPurchases } = itemDoc.data();
+	const { dateCreated, dateLastPurchased, dateNextPurchased, totalPurchases } =
+		itemDoc.data();
 
-	try {
-		return updateDoc(itemRef, {
-			dateLastPurchased: checked ? new Date() : dateLastPurchased,
-			totalPurchases: checked ? totalPurchases + 1 : totalPurchases,
-			checked: checked,
-		});
-	} catch (error) {
-		console.log('updateDoc error', error);
+	const checkedDateLastPurchased = checked ? new Date() : dateLastPurchased;
+
+	if (checkedDateLastPurchased) {
+		// Convert dateNextPurchased from Timestamp to a JS Date
+		const dateNextPurchasedAsDate = dateNextPurchased.toDate();
+
+		/**
+		 * Calculate the previous estimate for purchase
+		 * based on historical data
+		 */
+		const previousEstimate = getDaysBetweenDates(
+			checkedDateLastPurchased,
+			dateNextPurchasedAsDate,
+		);
+
+		/**
+		 * Calculate the days since the last transaction,
+		 * considering either last purchased or the created date
+		 */
+		const daysSinceLastTransaction = getDaysBetweenDates(
+			new Date(),
+			checkedDateLastPurchased || dateCreated,
+		);
+
+		// Calculate the remaining days until the next purchase
+		let remainingDays = calculateEstimate(
+			previousEstimate,
+			daysSinceLastTransaction,
+			totalPurchases,
+		);
+
+		// Ensure that the next purchase date is at least one day in the future
+		if (remainingDays <= 0) {
+			remainingDays = 1;
+		}
+
+		if (dateNextPurchasedAsDate < checkedDateLastPurchased) {
+			// Set checked to false so the item can be purchased again
+			checked = false;
+		}
+
+		try {
+			return updateDoc(itemRef, {
+				dateLastPurchased: checkedDateLastPurchased,
+				dateNextPurchased: getFutureDate(remainingDays),
+				totalPurchases: totalPurchases + 1,
+				checked: checked,
+			});
+		} catch (error) {
+			console.log('updateDoc error', error);
+		}
+	} else {
+		try {
+			/**
+			 * Prevent dateNextPurchased from being overridden
+			 * with the current date or dateCreated
+			 */
+			const newDateNextPurchased =
+				dateNextPurchased && dateNextPurchased.toDate();
+
+			return updateDoc(itemRef, {
+				dateLastPurchased: dateLastPurchased,
+				dateNextPurchased: newDateNextPurchased,
+				totalPurchases: totalPurchases,
+			});
+		} catch (error) {
+			console.log('Unchecked item update error:', error);
+		}
```

Comment:
Great work on this function! Useful comments and easy to follow.

Reply:
One non-critical suggestion (this is something we could do at the end): create a few helper functions (fetch the data, calculate dates, update Firestore) that feed into updateItem.

Reply:
I like that you're thinking along the lines of single responsibility!

Reply:
@CorinaMurg I love that idea! I'll give it a try today. If it gets too tricky, we should still do it later because that sounds great! 🥰 The thing is, we're calling two existing functions (one from `dates.js` and the other from the `@the-collab-lab/shopping-list-utils` module) and saving the results in variables. Because of that, I'm not sure how we can separate the 'calculate dates' part into a helper function, but I'll definitely give it a try and let you know. Thank you so much! ❤️